### PR TITLE
Add Nova Trustzone RSA keys support

### PR DESCRIPTION
**Pull request by isergieienkov**

### CODE REVIEW

File: src/nova_tz/keyblob.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        let start = 0;
+        let (magic_num, start) = read_u32_le(value, start);
+        let (version, start) = read_u32_le(value, start);
+        let (digest_padding, start) = read_u32_le(value, start);
+        let digest_padding = DigestPadAlgo::try_from(digest_padding)?;
+        let (modulus, start) = read_u8_array_dynamic(value, start, RSA_KEY_SIZE_MAX);
+        let (public_exponent, start) = read_u8_array_dynamic(value, start, RSA_KEY_SIZE_MAX);
+        let (iv, start) = read_u8_array(value, start, RSA_IV_LENGTH);
+        let (pvt_exponent, start) = read_u8_array_dynamic(value, start, RSA_KEY_SIZE_MAX);
+        let (hmac, _) = read_u8_array(value, start, RSA_HMAC_LENGTH);
```
Comment (JayKickliter) [inline src/nova_tz/keyblob.rs:82]: We can do away with manual index housekeeping with [`slice::split_at`]():

```suggestion
const RSA_KEY_BLOB_LENGHT: usize = 1656;

fn read_u32_le(buf: &[u8]) -> (u32, &[u8]) {
    let (val_bytes, rest) = buf.split_at(4);
    let parsed = u32::from_le_bytes(val_bytes.try_into().unwrap());
    (parsed, rest)
}

fn read_u8_array_dynamic(buf: &[u8], maxsize: usize) -> (Vec<u8>, &[u8]) {
    let (arr_bytes, rest) = buf.split_at(maxsize);
    let (arr_len, rest) = read_u32_le(rest);
    let arr = arr_bytes[..arr_len as usize].to_vec();
    (arr, rest)
}

fn read_u8_array(buf: &[u8], size: usize) -> (Vec<u8>, &[u8]) {
    let (arr_bytes, rest) = buf.split_at(size);
    let arr = arr_bytes.to_vec();
    (arr, rest)
}

impl TryFrom<&[u8]> for KeyBlob {
    type Error = Error;

    fn try_from(buf: &[u8]) -> Result<Self, Error> {
        // Key blobs are always exactly RSA_KEY_BLOB_LENGHT long, but
        // let's tolerate extra.
        if buf.len() < RSA_KEY_BLOB_LENGHT {
            return Err(Error::BadKeyBlob(format!(
                "invalid key blob len, expected: {RSA_KEY_BLOB_LENGHT}, got: {}",
                buf.len()
            )));
        }

        let (magic_num, rest) = read_u32_le(buf);
        let (version, rest) = read_u32_le(rest);
        let (digest_padding, rest) = read_u32_le(rest);
        let digest_padding = DigestPadAlgo::try_from(digest_padding)?;
        let (modulus, rest) = read_u8_array_dynamic(rest, RSA_KEY_SIZE_MAX);
        let (public_exponent, rest) = read_u8_array_dynamic(rest, RSA_KEY_SIZE_MAX);
        let (iv, rest) = read_u8_array(rest, RSA_IV_LENGTH);
        let (pvt_exponent, rest) = read_u8_array_dynamic(rest, RSA_KEY_SIZE_MAX);
        let (hmac, rest) = read_u8_array(rest, RSA_HMAC_LENGTH);

        // If the above parsing was correct, `rest` should be exactly
        // `RSA_KEY_BLOB_LENGHT` smaller than the original input
        // buffer.
        assert_eq!(buf.len() - RSA_KEY_BLOB_LENGHT, rest.len());

```

Comment (JayKickliter) [reply src/nova_tz/keyblob.rs:82]: we also need a test case to make sure that this parses a dummy key-blob

**Review by JayKickliter — CHANGES_REQUESTED** [review]

File: src/nova_tz/keyblob.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    pub version: u32,
+    pub digest_padding: DigestPadAlgo,
+    pub modulus: Vec<u8>,
+    pub public_exponent: Vec<u8>,
+    pub iv: Vec<u8>,
+    pub pvt_exponent: Vec<u8>,
+    pub hmac: Vec<u8>,
+}
+
+const RSA_KEY_SIZE_MAX: usize = 512 + 16;
```
[the remark this answers is not in the document]
Comment (isergieienkov) [reply src/nova_tz/keyblob.rs:45]: Don't know. Looks like some proprietary values from qualcomm

**Review by JayKickliter — DISMISSED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -30,12 +30,16 @@ byteorder = {version = "1", optional = true}
 multihash = {version = "0.18", optional = true}
 sqlx = {version = "0", optional = true, default-features=false}
 solana-sdk = {version = ">= 1, < 2", optional = true }
+num-bigint-dig = "0"
```
Comment (madninja) [inline Cargo.toml:33]: why are these three here? Are they really needed? If so they're only needed as part of nova-tz so make them `optional = true`

File: src/keypair.rs [hunk 4]
```diff
@@ -354,6 +375,20 @@ mod tests {
         sign_test_keypair(&Keypair::TPM(keypair));
     }
 
+    #[cfg(feature = "nova-tz")]
+    #[test]
+    fn sign_tz() {
+        fs::write(
+            "/tmp/rsa_key_blob",
```
Comment (madninja) [inline src/keypair.rs:382]: I suggest using https://crates.io/crates/tempfile as a dev-dependency to avoid file location/collision  issues and using it here

File: src/keypair.rs [hunk 5]
```diff
@@ -217,6 +236,8 @@ impl Deref for SharedSecret {
 mod tests {
     use super::*;
     use rand::rngs::OsRng;
+    #[cfg(feature = "tpm")]
```
Comment (madninja) [inline src/keypair.rs:239]: Why is this needed?

File: src/nova_tz/keyblob.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
+use std::convert::TryFrom;
+use std::convert::TryInto;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Can't parse key blob: {0}")]
+    BadKeyBlob(String),
```
[the remark this answers is not in the document]
Comment (madninja) [reply src/nova_tz/keyblob.rs:10]: I agree

File: Cargo.toml [hunk 7]
```diff
@@ -30,12 +30,16 @@ byteorder = {version = "1", optional = true}
...
+dynfmt = { version = "0", features = ["curly"] }
+uuid = { version = "1", features = ["v4", "fast-rng"] }
```
Comment (JayKickliter) [inline Cargo.toml:35]:
```suggestion
num-bigint-dig = { version = "0", optional = true }
dynfmt = { version = "0", features = ["curly"], optional = true }
uuid = { version = "1", features = ["v4", "fast-rng"], optional = true }
```

File: src/nova_tz/keyblob.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            2 => Ok(DigestPadAlgo::RsaPssSha2_256),
+            _ => Err(Error::BadKeyBlob(format!(
+                "Bad digest padding algorithm {}",
+                value
+            ))),
+        }
+    }
+}
+
+pub(crate) struct KeyBlob {
```
Comment (madninja) [inline src/nova_tz/keyblob.rs:35]: So technically a `KeyBlob` _is_ a `TrustZoneKeypair` or `TZKeypair` right? Why can't it be called that?

File: Cargo.toml [hunk 9]
```diff
@@ -30,12 +30,16 @@ byteorder = {version = "1", optional = true}
...
 
 [features]
 default = []
 ecc608 = [ "ecc608-linux" ]
 rsa = ["dep:rsa", "byteorder"]
 tpm = ["tss2", "libc", "drop_guard"]
+nova-tz = ["rsa"]
```
Comment (JayKickliter) [inline Cargo.toml:42]:
```suggestion
nova-tz = ["rsa", "num-bigint-dig", "dynfmt", "uuid"]
```

File: src/nova_tz/keyblob.rs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    pub public_exponent: Vec<u8>,
+}
+
+// Note (iegor.s): values are copied from qualcomm qseecom module
+const RSA_KEY_SIZE_MAX: usize = 512 + 16;
+const RSA_IV_LENGTH: usize = 16;
+const RSA_HMAC_LENGTH: usize = 32;
+const RSA_KEY_BLOB_LENGHT: usize = 1656;
+
+fn read_u32_le(buf: &[u8]) -> Result<(u32, &[u8]), TryFromSliceError> {
```
Comment (madninja) [inline src/nova_tz/keyblob.rs:46]: I do not believe you need these functions.. use a `Cursor` in the caller and the byteorder crate (which is already used in crypto-rs) for endianness readers on that cursor. That way you don't need these functions and just work with a mutable cursor on the slice of data

File: src/nova_tz/keyblob.rs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+impl TryFrom<&[u8]> for KeyBlob {
+    type Error = Error;
+
+    fn try_from(buf: &[u8]) -> Result<Self, Error> {
+        // Key blobs are always exactly RSA_KEY_BLOB_LENGHT long, but
+        // let's tolerate extra.
+        if buf.len() < RSA_KEY_BLOB_LENGHT {
+            return Err(Error::BadKeyBlob(format!(
```
Comment (madninja) [inline src/nova_tz/keyblob.rs:75]: Please rename KeyBlob to TZKeypair and then actually create a DecodeError enum for  each of the errors so we don't have all these strings for errors

File: src/nova_tz/mod.rs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+mod rsa_key;
+
+use crate::{public_key, rsa, KeyTag, KeyType, Network, Result, Sign};
+use ::rsa::RsaPublicKey;
+use std::fmt;
+
+extern crate num_bigint_dig as num_bigint;
+use num_bigint::BigUint;
+
+use crate::nova_tz::rsa_key::RsaKey;
```
Comment (madninja) [inline src/nova_tz/mod.rs:11]: no need for an absolute path to rsa_key in the use here right?

File: src/nova_tz/rsa_key.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    pub fn new(key_path: &str) -> Result<Self> {
+        let key_name = Uuid::new_v4().to_string();
+
+        if !Path::new(key_path).exists() {
+            return Err(Error::BadKeyPath(key_path.to_string()).into());
+        }
+
+        let key_blob_data =
+            fs::read(key_path).map_err(|_| Error::BadKeyPath(key_path.to_string()))?;
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:29]: Why mask io errors here?

File: src/nova_tz/rsa_key.rs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        let key_blob =
+            keyblob::KeyBlob::try_from(key_blob_data.as_ref()).map_err(Error::KeyBlobError)?;
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:31]:
```suggestion
            keyblob::KeyBlob::try_from(&key_blob_data).map_err(Error::KeyBlobError)?;
```

File: src/nova_tz/mod.rs [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("bad key path: {0}")]
+    BadKeyPath(String),
+
+    #[error("KeyBlob error")]
+    KeyBlobError(#[from] keyblob::Error),
+
+    #[error("qseecom error: {0}")]
+    QseecomError(String),
```
Comment (madninja) [inline src/nova_tz/mod.rs:23]: Why is this a string instead of the underlying error?

File: src/nova_tz/rsa_key.rs [hunk 16]
```diff
@@ -0,0 +1,75 @@
+use crate::nova_tz::keyblob;
+use crate::nova_tz::Error;
+use crate::nova_tz::Result;
+use dynfmt::{Format, SimpleCurlyFormat};
+use std::convert::TryFrom;
+use std::fs;
+use std::path::Path;
+use uuid::Uuid;
+
+pub struct RsaKey {
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:10]: This name confused me a lot.. it's not an RSAKey but, what is it? A TZKeypair registry?

File: src/nova_tz/rsa_key.rs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        Ok((
+            self.key_blob.modulus.clone(),
+            self.key_blob.public_exponent.clone(),
+        ))
+    }
+
+    pub fn sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
+        let sign_path = SimpleCurlyFormat
+            .format(RsaKey::NOVA_RSA_SIGN_PATH, [&self.key_name])
+            .unwrap()
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:57]: don't unwrap?

File: src/nova_tz/rsa_key.rs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+                .unwrap()
+                .to_string(),
+            key_blob_data,
+        )
+        .map_err(|e| Error::QseecomError(e.to_string()))?;
+
+        Ok(RsaKey { key_name, key_blob })
+    }
+
+    pub fn public_key(&self) -> Result<(Vec<u8>, Vec<u8>)> {
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:47]: * Why not return a `crate::PublicKey` instead of a tuple?
* Why does it need to be cloned?

Perhaps this just needs a `implFrom<RsaKey> for crate::PublicKey`

**Review by madninja — CHANGES_REQUESTED** [review]

File: src/nova_tz/keyblob.rs (hunk 10 above)
Comment (isergieienkov) [reply src/nova_tz/keyblob.rs:46]: It will lead to ugly repeatable code. because I have to store the buffer of full size anyway until I read real size of it.

File: src/nova_tz/keyblob.rs (hunk 11 above)
Comment (isergieienkov) [reply src/nova_tz/keyblob.rs]: Not sure if I understood your request about errors correctly. If it's not what you meant - please describe.

File: src/nova_tz/keyblob.rs (hunk 10 above)
Comment (madninja) [reply src/nova_tz/keyblob.rs:64]: I'm not sure that's the case.. Do you have a test with a binary that gets decoded? I can try it to see if I can prove you wrong

Comment (JayKickliter) [reply src/nova_tz/keyblob.rs:64]: he is correct about reading exponents (see `read_u8_array_dynamic`), because they are length tagged, but the Len tag is after the buffer. So you do need to read the whole buffer first. What you're seeing here includes my "improvements", which essentially re-implemnet a couple pieces of `nom`. 

I also saw other ways to improve this, but I also knew he's against a huge time crunch.

Comment (JayKickliter) [reply src/nova_tz/keyblob.rs:64]: the way to do this with pure std+cursor is read max sized dynamic buffer into a vector, and resize it after reading the following length tag

Comment (madninja) [reply src/nova_tz/keyblob.rs:64]: Right.. that's what I thought too.. It'll remove all these functions. I don't think that work is critical though so we can live with the implementation here for a bit while we get through the time crunch

File: src/nova_tz/rsa_key.rs [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+            "/sign"
+        );
+
+        fs::write(&sign_path, msg).map_err(Error::QseecomError)?;
+        let signed = fs::read(&sign_path).map_err(Error::QseecomError)?;
+        Ok(signed)
+    }
+
+    fn remove_key(&self) -> Result {
+        fs::write(TzRsaKeyInfo::NOVA_RSA_REMOVE_PATH, &self.key_name).unwrap();
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:64]: why unwrap here? Are we trying to ignore the error? or actually crash?

File: src/nova_tz/rsa_key.rs (hunk 13 above)
Comment (madninja) [reply src/nova_tz/rsa_key.rs]: I'm still not a fan of ignoring the io::Error and converting it into a string here.

**Review by madninja — COMMENTED** [review]
Couple of updated comments as part of the review

File: src/nova_tz/rsa_key.rs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    key_name: String,
+    rsa_key: RsaPublicKey,
+}
+
+impl TzRsaKeyInfo {
+    const NOVA_RSA_KEYS_PATH: &'static str = "/sys/nova-rsa/keys/";
+    const NOVA_RSA_ADD_PATH: &'static str = "/sys/nova-rsa/add";
+    const NOVA_RSA_REMOVE_PATH: &'static str = "/sys/nova-rsa/remove";
+
+    pub fn from_path(key_path: &str) -> Result<Self> {
```
Comment (madninja) [inline src/nova_tz/rsa_key.rs:19]: Why not take an actual `key_path: &Path` instead of a string?

Comment (isergieienkov) [reply src/nova_tz/rsa_key.rs:19]: How this will help? `exists` is the only functionality needed from Path.

File: src/nova_tz/keyblob.rs (hunk 10 above)
Comment (isergieienkov) [reply src/nova_tz/keyblob.rs]: I've reimplemented it using Cursor. But I also want to remove all that errors (except for 3 field length checks). Because there is total length check at the beginning it's impossible to get Cursor reading errors, right?

**Comment by madninja** [comment]
They’re all io::Error right? You can create those

**Comment by isergieienkov** [comment]
Yes. What do you mean create? `data.read` and `data.read_u32` shouldn't throw any errors because it's guaranteed that there is enough data in the buffer. Am I correct? I want to keep only errors from checks that length read is not greater that maximum allowed.

File: src/nova_tz/rsa_key.rs (hunk 20 above)
Comment (madninja) [reply src/nova_tz/rsa_key.rs:19]: Right and Path is he type of thing you're looking for. Better/clearer types..

**Comment by madninja** [comment]
> Yes. What do you mean create? `data.read` and `data.read_u32` shouldn't throw any errors because it's guaranteed that there is enough data in the buffer. Am I correct? I want to keep only errors from checks that length read is not greater that maximum allowed.

But you have a number of `map_err` after a read/read_u32 in the PR right? Each of those is not passing the underlying error back up

**Review by madninja — APPROVED** [review]
